### PR TITLE
Massive changes.

### PR DESCRIPTION
**Pull request by mcb64**
All output records are passive scanned, all input records are I/O Intr scanned.

The EK9000_PollThread has been changed to read *all* of the input data at 5Hz, and reset the watchdog at 2.5Hz.

The devEK9000Terminal class has a new getEK9000IO routine to read the input previously buffered by EK9000_PollThread.
  
The EL{1,3,5}XXX are all input devices, and their device support now has get_ioint_info functions to allow I/O Intr scanning.  The read routines now directly call the "callback" routines, as callbacks are no longer needed as I/O has been completed before record processing starts.

The output devices have been changed to correctly handle PACT. (PACT needs to be set in the read routine and the callback needs to reprocess the record after reading the data.  The reprocessing should clear PACT.)

Removed extra push_back of devEK9000 to master device list.

Fixed EL5042 input length in terminals.h and read length in devEL5XXX.cpp.

Don't ignore local configs... we need to check them in *somewhere*, if only locally!

Removed setupReadCallback, as we don't do this any more.  And even if we did, we don't want to clear PACT.

### CODE REVIEW

**Comment by mcb64** [comment]
@klauer or @zllentz should probably be added as reviewers here, but I don't seem to be able to do this.

**Comment by JJL772** [comment]
For some reason I can't request review from @ZLLentz, so just leaving it with me and Ken for now.

**Comment by ZLLentz** [comment]
Bizarre. I'll still take a look.

File: ek9000App/src/devEL50XX.cpp
Comment (JJL772) [inline ek9000App/src/devEL50XX.cpp]: (nitpick) The contents of el5042_read_callback can probably just be moved here.

File: ek9000App/src/devEL50XX.cpp
Comment (JJL772) [inline ek9000App/src/devEL50XX.cpp]: (nitpick) The contents of el50xx_read_callback can probably just be moved here.

File: ek9000App/src/devEL3XXX.cpp
Comment (JJL772) [inline ek9000App/src/devEL3XXX.cpp]: (nitpick) The contents of EL30XX_ReadCallback can be moved to this procedure.

File: ek9000App/src/devEL1XXX.cpp
Comment (JJL772) [inline ek9000App/src/devEL1XXX.cpp]: (nitpick) EL10XX_ReadCallback can be moved into this function

File: ek9000App/src/devEK9000.cpp
Comment (JJL772) [inline ek9000App/src/devEK9000.cpp]: Since this isn't guarded by a mutex, isn't it possible that a record could read from `m_analog_buf` or `m_digital_buf` while those register spaces are filled with new data by the poll function? A read-write mutex could be used as a low-cost guard around this, although I don't think EPICS has one of those.

**Review by JJL772 — COMMENTED** [review]
Looks pretty good to me, just some minor questions/feedback.

There's also a terminals.json file that is used to automatically generate the terminals.h header. The input PDO size for the EL5042 should be updated there too, however I can do that post-merge (or by pushing commits to your fork, with permission)

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [inline ek9000App/src/devEK9000.cpp:104]: Nitpicky: `cnt` is really `should_check_connectivity` as it only goes 0<->1

More nitpicky: looks like you need to run `clang-format` in the repo root

**Comment by klauer** [comment]
Am taking a look through this now, but it'll take me a bit. (Sorry for the delay, I'm not too familiar with the codebase and have been off in Python/PLC land for much too long...)

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [inline ek9000App/src/devEK9000.cpp]: `status != 0` means we failed to acquire the lock above (`asynError`); should `Unlock` even be here? ... or for that matter, should we even be executing this block?

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [inline ek9000App/src/devEK9000.cpp:112]: Why isn't `status` checked right here? More on that below; the logic may be off (or maybe it's just Monday...)

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [inline ek9000App/src/devEK9000.cpp]: Hmm, I don't see how this failure to read can end up resulting in `READ/INVALID` being set on a per-record basis. 
Poll thread -> `m_digital_buf` -> `getEK9000IO` gets stale data

Am I misunderstanding the flow of data, or missing something else here?

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [inline ek9000App/src/devEK9000.cpp:151]: (Just a note, no action required: `gettimeofday` is susceptible to the system clock changing based on NTP and other things, but the `if (duration < poll delay)` check here makes that not a problem anyway; at worst we poll once a bit faster or slower on the change.)

File: ek9000App/src/devEL50XX.cpp
Comment (mcb64) [reply ek9000App/src/devEL50XX.cpp]: Done, pushing shortly.

File: ek9000App/src/devEL3XXX.cpp
Comment (mcb64) [reply ek9000App/src/devEL3XXX.cpp]: Done, pushing shortly.

File: ek9000App/src/devEL1XXX.cpp
Comment (mcb64) [reply ek9000App/src/devEL1XXX.cpp]: Done, pushing shortly.

File: ek9000App/src/devEK9000.cpp
Comment (mcb64) [reply ek9000App/src/devEK9000.cpp]: Well, we have Lock/Unlock defined, might as well use them!  Pushing a change shortly.

**Review by JJL772 — APPROVED** [review]
All of my complaints (and nitpicks) have been resolved, looks great otherwise! We'll still wait for approval from Ken & Zach before merging.

File: ek9000App/src/devEL2XXX.cpp
Comment (klauer) [inline ek9000App/src/devEL2XXX.cpp]: Isn't `device->Lock` going to `asynPortDriver::lock` which returns `asynSuccess` or `asynError`? (Fortunately both success values are equal, in either case...)

**Review by klauer — APPROVED** [review]
I had to read through the codebase to get some perspective here - and I've learned a lot about how all this works. I think the changes overall look really good and offer some big improvements to what's already a great EPICS support module. Nice work, Jeremy and Mike!

I found some minor things and made some notes about them above.

We may also want to give EED / the accelerator (ping @mamontironi) a heads-up that this big change is on the way.

File: ek9000App/src/devEK9000.cpp
Comment (JJL772) [reply ek9000App/src/devEK9000.cpp:104]: I'm going to run `clang-format` again after this merges I think.

File: ek9000App/src/devEL2XXX.cpp
Comment (JJL772) [reply ek9000App/src/devEL2XXX.cpp]: Good catch- originally it was using a standard epics mutex, but I changed that sometime in 2019 iirc (gosh time flies)

File: ek9000App/src/devEK9000.cpp
Comment (mcb64) [reply ek9000App/src/devEK9000.cpp:104]: Why this is cnt: originally, I was trying to run at 10 Hz, and do the connectivity change every third iteration.  Once I moved to 5Hz, I decided every other iteration was better.

We might still move to 10Hz, so I'd prefer to leave cnt for now.

And yeah, apologies for the inconsistent indents. ;-)

File: ek9000App/src/devEK9000.cpp
Comment (mcb64) [reply ek9000App/src/devEK9000.cpp]: Point taken, if we can't lock, we should just immediately continue the loop.  (I'm not convinced that Lock can fail in the absence of memory corruption, but... )

Change to be pushed shortly.

File: ek9000App/src/devEK9000.cpp
Comment (mcb64) [reply ek9000App/src/devEK9000.cpp]: We don't scan unless the read succeeds.  So if the read fails, the record just doesn't process.

So the data will be old, but so will the timestamp... the record doesn't update.

If we wanted to put out a read error, we could add {digital/analog}_failed flags here, scan in any case, and have getEK9000IO fail if the flag is set.

I'm open to suggestions on exactly what behavior we want here.
      1. Silent failure (keeping old data/timestamp without record processing).
      2. Setting a read error with the old data.
      3. Setting a read error with zero (or some other fixed value).

File: ek9000App/src/devEL2XXX.cpp
Comment (mcb64) [reply ek9000App/src/devEL2XXX.cpp]: Fixed, push forthcoming.

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [reply ek9000App/src/devEK9000.cpp]: Ah, good point about the processing. I missed that important step.

So the current behavior is (1).

I personally think (2) is preferable - alarm sensitive clients will know there's something wrong, but non-alarm sensitive clients won't be thrown by a `0` value as could happen with (3). That is, there are so many scripts that are just `epics.caget(pvname)` and throw away alarm values entirely. A stale value is probably better than no value, in that case.

Other opinions, @JJL772 @ZLLentz @mamontironi?

Comment (mcb64) [reply ek9000App/src/devEK9000.cpp]: I think you're right, actually.  Change pushed.

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [reply ek9000App/src/devEK9000.cpp:104]: Fair enough!

File: ek9000App/src/devEK9000.cpp
Comment (klauer) [reply ek9000App/src/devEK9000.cpp]: Changes look good. Thanks!

**Comment by klauer** [comment]
Nothing else to add on my end. Looks good 👍

**Review by JJL772 — APPROVED** [review]

**Comment by JJL772** [comment]
Would definitely like to get this in soon. 
Is @ZLLentz still going to review this?

**Comment by mcb64** [comment]
I think I'm OK with you and Ken looking over it, even if Zach doesn't.

**Comment by ZLLentz** [comment]
I don't think I'm going to get around to reviewing it- sorry about that

**Pull request merged**
